File: src/main/java/org/openbase/planetsudo/geometry/BoundingBox.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package org.openbase.planetsudo.geometry;

/*-
 * #%L
 * PlanetSudo GameEngine
 * %%
 * Copyright (C) 2009 - 2019 openbase.org
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

/**
 *
 * @author <a href="mailto:divine@openbase.org">Divine Threepwood</a>
 */
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics2D;
import java.awt.Point;
import java.awt.geom.AffineTransform;
public class BoundingBox {

	private int maxX, maxY, minX, minY;

	public BoundingBox() {
		this.maxX = Integer.MIN_VALUE;
		this.maxY = Integer.MIN_VALUE;
		this.minX = Integer.MAX_VALUE;
		this.minY = Integer.MAX_VALUE;
	}

	public int getMaxX() {
		return maxX;
	}

	public int getMaxY() {
		return maxY;
	}

	public int getMinX() {
		return minX;
	}

	public int getMinY() {
		return minY;
	}

	public int getCenterX() {
		return minX + ((maxX - minX) / 2);
	}

	public int getCenterY() {
		return minY + ((maxY - minY) / 2);
	}

	public Point getCenter() {
		return new Point(getCenterX(),getCenterY());
	}

	public int getHeight() {
		return (checkYValidation() ? maxY - minY : 0);
	}
	public int getWidth() {
		return (checkXValidation() ? maxX - minX : 0);
	}

	public void insertX(int x) {
		minX = Math.min(minX, x);
		maxX = Math.max(maxX, x);
	}

	public void insertY(int y) {
		minY = Math.min(minY, y);
		maxY = Math.max(maxY, y);
	}

	public void insertBoundingBox(BoundingBox boundingBox) {
		this.maxX = Math.max(maxX,boundingBox.getMaxX());
		this.maxY = Math.max(maxY,boundingBox.getMaxY());
		this.minX = Math.min(minX,boundingBox.getMinX());
		this.minY = Math.min(minY,boundingBox.getMinY());
	}

	public boolean checkXValidation() {
		return minX < maxX;
	}

	public boolean checkYValidation() {
		return minY < maxY;
	}

	public void paint(Graphics2D g, AffineTransform trans) {
		if(!(checkXValidation() || checkYValidation())) {
			return;
		}

		Graphics2D g2 = (Graphics2D) g.create();
		g2.transform(trans);
		g2.setColor(Color.MAGENTA);
		g2.drawRect(minX, minY, maxX - minX, maxY - minY);
		g2.setColor(Color.ORANGE);
		g2.drawLine(getCenterX(), getCenterY(), minX, minY);
		g2.drawLine(getCenterX(), getCenterY(), maxX, maxY);
		g2.drawLine(getCenterX(), getCenterY(), maxX, minY);
		g2.drawLine(getCenterX(), getCenterY(), minX, maxY);
		g2.dispose();
	}
	
	public Dimension getDimension() {
		return new Dimension(getWidth(), getHeight());
	}
}

File: src/main/kotlin/org/openbase/planetsudo/geometry/BoundingBox.kt
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package org.openbase.planetsudo.geometry

import java.awt.Color
import java.awt.Dimension
import java.awt.Graphics2D
import java.awt.Point
import java.awt.geom.AffineTransform
import kotlin.math.max
import kotlin.math.min

/**
 *
 * @author [Divine Threepwood](mailto:divine@openbase.org)
 */

class BoundingBox {
    var maxX: Int
        private set
    var maxY: Int
        private set
    var minX: Int
        private set
    var minY: Int
        private set

    init {
        this.maxX = Int.MIN_VALUE
        this.maxY = Int.MIN_VALUE
        this.minX = Int.MAX_VALUE
        this.minY = Int.MAX_VALUE
    }

    val centerX: Int
        get() = minX + ((maxX - minX) / 2)

    val centerY: Int
        get() = minY + ((maxY - minY) / 2)

    val center: Point
        get() = Point(centerX, centerY)

    val height: Int
        get() = (if (checkYValidation()) maxY - minY else 0)
    val width: Int
        get() = (if (checkXValidation()) maxX - minX else 0)

    fun insertX(x: Int) {
        minX = min(minX.toDouble(), x.toDouble()).toInt()
        maxX = max(maxX.toDouble(), x.toDouble()).toInt()
    }

    fun insertY(y: Int) {
        minY = min(minY.toDouble(), y.toDouble()).toInt()
        maxY = max(maxY.toDouble(), y.toDouble()).toInt()
    }

    fun insertBoundingBox(boundingBox: BoundingBox) {
        this.maxX = max(maxX.toDouble(), boundingBox.maxX.toDouble()).toInt()
        this.maxY = max(maxY.toDouble(), boundingBox.maxY.toDouble()).toInt()
        this.minX = min(minX.toDouble(), boundingBox.minX.toDouble()).toInt()
        this.minY = min(minY.toDouble(), boundingBox.minY.toDouble()).toInt()
    }

    fun checkXValidation(): Boolean {
        return minX < maxX
    }

    fun checkYValidation(): Boolean {
        return minY < maxY
    }

    fun paint(g: Graphics2D, trans: AffineTransform?) {
        if (!(checkXValidation() || checkYValidation())) {
            return
        }

        val g2 = g.create() as Graphics2D
        g2.transform(trans)
        g2.color = Color.MAGENTA
        g2.drawRect(minX, minY, maxX - minX, maxY - minY)
        g2.color = Color.ORANGE
        g2.drawLine(centerX, centerY, minX, minY)
        g2.drawLine(centerX, centerY, maxX, maxY)
        g2.drawLine(centerX, centerY, maxX, minY)
        g2.drawLine(centerX, centerY, minX, maxY)
        g2.dispose()
    }

    val dimension: Dimension
        get() = Dimension(width, height)
}
